app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt

```diff
@@ -0,0 +1,101 @@
+package github.com.rhacco.dota2androidapp.activities
+
+import android.os.Bundle
+import android.support.v7.app.AppCompatActivity
+import android.util.Log
+import github.com.rhacco.dota2androidapp.R
+import github.com.rhacco.dota2androidapp.lists.LiveMatchesListAdapter
+import github.com.rhacco.dota2androidapp.sources.remote.getDota2OfficialAPIService
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.schedulers.Schedulers
+import kotlinx.android.synthetic.main.activity_live_matches.*
+
+class LiveMatchesActivity : AppCompatActivity() {
+    private lateinit var mListAdapter: LiveMatchesListAdapter
+    private val mDisposables = CompositeDisposable()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_live_matches)
+        actionBar?.setDisplayHomeAsUpEnabled(true)
+        mListAdapter = LiveMatchesListAdapter(this)
+        listView.adapter = mListAdapter
+        updateLiveMatches()
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        mDisposables.clear()
+    }
+
+    private fun updateLiveMatches() {
+        val disposable = getDota2OfficialAPIService()
+                ?.fetchTopLiveGames(getString(R.string.api_key), 0)
+                ?.subscribeOn(Schedulers.io())
+                ?.observeOn(AndroidSchedulers.mainThread())
+                ?.subscribe(
+                        { result ->
+                            for (game in result.game_list)
+                                updateListEntry(game.server_steam_id, game.average_mmr)
+                        },
+                        { error ->
+                            Log.d(getString(R.string.log_msg_debug),
+                                    "Failed to update top live games: " + error)
+                        }
+                )
+        if (disposable != null)
+            mDisposables.add(disposable)
+    }
+
+    private fun updateListEntry(serverSteamId: Long, averageMMR: Int) {
+        val disposable = getDota2OfficialAPIService()
```

Comment:
Same as above here.

Any advanced manipulation of the format of the response should be done on the repository level. By advanced manipulation, I mean text formatting of the response or the conversion of the response to another data structure (ie array to LinkedList) or something.

The activity should be more or less responsible for seeing a list of objects and throwing that to the adapter.

Databinding with the XML can also be used to display a certain string(s) in a particular order. Consider:


<layout>
<data>
        <variable
            name="match"
            type="foo.bar.Match" />
</data>

<TextView
android:id="@+id/player0
android:src="@{match.player[0]}"
/>
....
</layout>

String manipulation in the activity is entirely inappropriate.

Reply:
So from what you've told me I thought the ViewModel > Repository architecture only applies to queries with a database in the background since you said DAO + API = Repository. I don't use a DAO/database here since live data changes frequently, only the direct query on the API, so I considered the extra classes superfluous and got rid of them.

Right now I'm seeing "The ViewModel class is designed to store and manage UI-related data so that the data survives configuration changes such as screen rotations" in official Android docs so I definitely see the point of introducing this architecture again. I think storing the live data in memory in a new LocalDataSource is the way to go (still no database but with the ViewModel architecture), so I'm going with this now and let you take another look at it